### PR TITLE
Replace ai.x.diff library with com.softwaremill.diffx [DPP-192]

### PR DESCRIPTION
**Pull request by kamil-da**
**Motivation**
At present the Ledger API Test Tool relies on `ai.x.diff` for capturing difference between case classes. This project is no longer maintained and there is no version supporting scala 2.13. We have to switch to using another library so that a full upgrade to scala 2.13 can be performed.

**Changes**
Replaced the `ai.x.diff` library by `diffx` by SoftwareMill.
[https://github.com/softwaremill/diffx](https://github.com/softwaremill/diffx)

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-test-tool/BUILD.bazel [hunk 1]
```diff
@@ -148,8 +148,10 @@ EOF
                 "//ledger/test-common:semantic-tests%s.scala" % lf_version_suffix,
                 "//libs-scala/grpc-utils",
                 "//libs-scala/timer-utils",
-                "@maven//:ai_x_diff_2_12",
                 "@maven//:com_chuusai_shapeless_2_12",
+                "@maven//:com_propensive_magnolia_2_12",
+                "@maven//:com_propensive_mercator_2_12",
```
Comment (kamil-da) [inline ledger/ledger-api-test-tool/BUILD.bazel:153]: While these libraries are not directly used in our code, they are transitive dependencies of the `diffx`. Without them being explicitly listed Bazel will not put them on the classpath and automatic type class derivation will not work.

File: bazel-java-deps.bzl [hunk 2]
```diff
@@ -173,6 +173,7 @@ def install_java_deps():
             "io.gatling:gatling-http-client:3.3.1",
             "com.fasterxml.jackson.core:jackson-core:2.12.0",
             "com.fasterxml.jackson.core:jackson-databind:2.12.0",
+            "com.softwaremill.diffx:diffx-core_{}:0.4.1".format(scala_major_version),
```
Comment (cocreature) [inline bazel-java-deps.bzl:176]: can we kill the ai.x.diff dependency now?

File: ledger/ledger-api-test-tool/BUILD.bazel (hunk 1 above)
Comment (SamirTalwar-DA) [reply ledger/ledger-api-test-tool/BUILD.bazel:153]: Perhaps add a comment to the Bazel file as such?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/DiffExtensions.scala [hunk 3]
```diff
@@ -3,51 +3,45 @@
...
+      }
 
-  implicit def diffShowSeq[T](implicit diffShowT: DiffShow[T]): DiffShow[Seq[T]] =
-    new DiffShow[Seq[T]] {
-      override def show(t: Seq[T]): String = t.toString
+    }
+  }
 
-      override def diff(left: Seq[T], right: Seq[T]): Comparison = {
+  implicit def diffSeq[T](implicit diffT: Diff[T]): Diff[Seq[T]] =
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/DiffExtensions.scala:20]: So the new library doesn't have a native `Diff[Seq[T]]` either?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/DiffExtensions.scala [hunk 4]
```diff
@@ -3,51 +3,45 @@
...
+    new Diff[Seq[T]] {
+      override def apply(left: Seq[T], right: Seq[T], toIgnore: List[FieldPath]): DiffResult = {
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/DiffExtensions.scala:22]: You seem to be ignoring `toIgnore`.

File: maven_install_2.12.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
     "dependency_tree": {
-        "__AUTOGENERATED_FILE_DO_NOT_MODIFY_THIS_FILE_MANUALLY": 1485856285,
+        "__AUTOGENERATED_FILE_DO_NOT_MODIFY_THIS_FILE_MANUALLY": 1031077523,
```
Comment (cocreature) [inline maven_install_2.12.json:3]: you also need to repin the `2.13` lockfile.

File: ledger/ledger-api-test-tool/BUILD.bazel (hunk 1 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/BUILD.bazel:153]: I think at this point everybody working with Bazel is already familiar with the fact that transitive dependencies must be listed explicitly.

File: bazel-java-deps.bzl (hunk 2 above)
Comment (stefanobaghino-da) [reply bazel-java-deps.bzl:176]: kill it with fire

File: ledger/ledger-api-test-tool/BUILD.bazel (hunk 1 above)
Comment (cocreature) [reply ledger/ledger-api-test-tool/BUILD.bazel:153]: the errors here look pretty different from the ones you get usually so I think it’s worth adding a comment even for people familiar with Bazel.

Comment (SamirTalwar-DA) [reply ledger/ledger-api-test-tool/BUILD.bazel:153]: That, and so I know exactly what's depending on them. Something like `// transitive dependency of diffx` would help me a lot.

Comment (kamil-da) [reply ledger/ledger-api-test-tool/BUILD.bazel]: I've added comments for both `magnolia` and `mercator`. As @cocreature mentioned errors in this case were rather misleading, especially for the `mercator`.

File: bazel-java-deps.bzl (hunk 2 above)
Comment (kamil-da) [reply bazel-java-deps.bzl:174]: Done 🚮

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/DiffExtensions.scala (hunk 3 above)
Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/DiffExtensions.scala]: Good catch! There's already available instance of `Diff[Seq[T]]`, so I've removed the custom one.

File: maven_install_2.12.json (hunk 5 above)
Comment (kamil-da) [reply maven_install_2.12.json]: Done.

**Review by cocreature — APPROVED** [review]
Awesome, thanks! Do you have an example of a diff we got before and the diff we get now?

**Comment by kamil-da** [comment]
> Awesome, thanks! Do you have an example of a diff we got before and the diff we get now?

Sure, see here:

**Simple status comparison**
`ai.x.diff`:
```
  SERVING -> UNKNOWN
```

`com.softwaremill.diffx`:
```
  io.grpc.health.v1.health.HealthCheckResponse.ServingStatus.SERVING -> io.grpc.health.v1.health.HealthCheckResponse.ServingStatus.UNKNOWN
```

**Objects comparison**
`ai.x.diff`:
```
  Seq(
    RecordField( ...,
      value = Some( Value( Text( "CKExposedByTemplate-alpha-150ec2f0cb212-key-0" -> "CKExposedByTemplate-alpha-150ec2f0cb212-key-1" ) ) )
    )
  )
```

`com.softwaremill.diffx`:
```
  List(
       0: RecordField(_1,Some(Value(Party(CKExposedByTemplate-alpha-1515516bb823d-party-0)))),
       1: RecordField(
            label: _2,
            value: Some(
                 value: Value(
                      sum: Text(
                           value: CKExposedByTemplate-alpha-1515516bb823d-key-0 -> CKExposedByTemplate-alpha-1515516bb823d-key-1)))))
```

**Sequence comparison (additional elements)**
`ai.x.diff` with our in-house instance of `DiffShow[Seq[T]]`:
```
  Seq( ...,
    2 = RecordField( label = "giver", value = Some( Value( Party( "CSDiscloseCreateToObservers-alpha-150ec2f0cb212-party-0" ) ) ) ),
    3 = RecordField(
      label = "observers",
      value = Some( Value( List( List( Vector(Value(Party(CSDiscloseCreateToObservers-alpha-150ec2f0cb212-party-1)), Value(Party(CSDiscloseCreateToObservers-alpha-150ec2f0cb212-party-2))) ) ) ) )
    )
  )
```

`com.softwaremill.diffx`:
```
  List(
       0: RecordField(giver,Some(Value(Party(CSDiscloseCreateToObservers-alpha-1515516bb823d-party-0)))),
       1: RecordField(observers,Some(Value(List(List(Vector(Value(Party(CSDiscloseCreateToObservers-alpha-1515516bb823d-party-1)), Value(Party(CSDiscloseCreateToObservers-alpha-1515516bb823d-party-2)))))))),
       2: RecordField(giver,Some(Value(Party(CSDiscloseCreateToObservers-alpha-1515516bb823d-party-0)))),
       3: RecordField(observers,Some(Value(List(List(Vector(Value(Party(CSDiscloseCreateToObservers-alpha-1515516bb823d-party-1)), Value(Party(CSDiscloseCreateToObservers-alpha-1515516bb823d-party-2)))))))))
```

I would say that `diffx` has nicer diffs. The only problematic case might be for comparing sequences in monochrome terminals because the `diffx` uses colours to differentiate additional/missing elements in their default `Diff[Iterable]` instance. Our custom instance for `DiffShow[Seq[T]]` had such problem solved by not only printing values but also their corresponding indices.
@cocreature Do you think this may be an issue for us?

**Comment by kamil-da** [comment]
A screenshot for more colorful experience:
Screenshot 2021-02-15 at 18 02 26

**Comment by cocreature** [comment]
Being harder to read if you don’t have colors is fine imho. Being unusable not, e.g., in CI environments it’s not that uncommon for colors to not work properly. Does the library have a way to detect if you’re writing to a terminal and  disable colors?

**Comment by kamil-da** [comment]
I cannot see an option to disable colours in favour of some other graphical differentiation.
A solution would be to leave our custom instance for iterables.

**Comment by cocreature** [comment]
That seems slightly disappointing. I’ll leave it to you and your team to decide whether that’s worth addressing or not.

**Comment by stefanobaghino-da** [comment]
> That seems slightly disappointing. I’ll leave it to you and your team to decide whether that’s worth addressing or not.

Given that this was the situation we had before, I'd say it's fine to leave it like that.

**Comment by kamil-da** [comment]
> > That seems slightly disappointing. I’ll leave it to you and your team to decide whether that’s worth addressing or not.
> 
> Given that this was the situation we had before, I'd say it's fine to leave it like that.

@stefanobaghino-da  I'm not sure if I understand correctly - you'd vote for leaving the default behaviour of `diffx` for iterables which is printing diffs in colours rather than explicitly with index annotation as we had before?

**Comment by stefanobaghino-da** [comment]
> > > That seems slightly disappointing. I’ll leave it to you and your team to decide whether that’s worth addressing or not.
> > 
> > 
> > Given that this was the situation we had before, I'd say it's fine to leave it like that.
> 
> @stefanobaghino-da I'm not sure if I understand correctly - you'd vote for leaving the default behaviour of `diffx` for iterables which is printing diffs in colours rather than explicitly with index annotation as we had before?

I was just saying the differentiation by color is what we already had before, so it should be fine to keep it like that. That means, I don't think there's a need to work on additional differentiation.

**Comment by kamil-da** [comment]
@stefanobaghino-da Ok 👍  Thank you

**Comment by cocreature** [comment]
:shipit:

**Pull request merged**